packages/box/src/Box.tsx

```diff
@@ -1,68 +1,69 @@
 import React from 'react';
 import PropTypes from 'prop-types';
-import { HTMLElementProps } from '@leafygreen-ui/lib';
-import omit from 'lodash/omit';
 
-type DivProps<T> = HTMLElementProps<'div'> & T;
+type InferComponentType<C, H> = H extends string
+  ? 'a' | React.ComponentType
+  : C;
 
-type AnchorProps<T> = HTMLElementProps<'a'> &
-  T & {
-    href: string;
-  };
-
-type CustomElementProps<T> = T & {
-  component: React.ElementType<any>;
-  [key: string]: any;
-};
-
-export type BoxProps<T> = DivProps<T> | AnchorProps<T> | CustomElementProps<T>;
-
-function isCustomElement<T>(
-  props: BoxProps<T>,
-): props is CustomElementProps<T> {
-  return (props as any).component != null;
+interface BoxComponentProps<
```

Comment:
This interface can be removed by changing `BoxProps` to:
```ts
export type BoxProps<
  C extends React.ElementType,
  H extends string | undefined,
  P extends Record<string, any> = {}
> = P & {
  href?: H;
  as?: InferComponentType<C, H>;
} & Omit<
  React.ComponentPropsWithRef<C extends React.ComponentType ? any : C>,
  'href' | 'as'
>;
```

Generally, I'm just seeing a lot of repeating type parameters and such, so this can help trim that down a bit.

Reply:
👍